### PR TITLE
Task/83 schema changes round2

### PR DESCRIPTION
**Pull request by rg911**
### Schema changes 

- ModelSchema
  1. `chainInfo` -> `chainStatistic`
  2. `signer` -> `signerPublicKey`
  3. `blockTransactionsHash` -> `transactionsHash`
  4. `blockReceiptsHash` -> `receiptsHash`
  5. `beneficiary` -> `beneficiaryPublicKey`
  
- ModelDescriptor
  1. MosaicDescriptor: `mosaicId` -> `id, Added: [`startHeight`, `ownerPublicKey`, `ownerAddress`, `revision`]
  2. MosaicDefinition.mosaicProperty: Restructured the schema (array -> object). Added [`duration`]
  3. HashLockInfo: `account` -> `senderPublicKey`, `accountAddress` -> `senderAddress`, `mosaic` -> 'mosaicId`, Added `amount`, `height` -> `endHeight`
  4. SecretLockInfo: `account` -> `senderPublicKey`, `accountAddress` -> `senderAddress`, `mosaic` -> `mosaicId`, Added `amount`, `height` -> `endHeight`, Added `compositeHash`
  5. Multisig: `account` -> `accountPublicKey`, `accountAddress` -> `address`
  6. Namespace: `owner` -> `ownerPublicKey`
  7. Receipt: BalanceChange.account -> `targetPublicKey`, BalanceTransfer.sender -> `senderPublicKey`, BalanceTransfer.recipient -> `recipientAddress`

- Transactions
  1. AccountLink: `remoteAccountKey` -> `remotePublicKey`, 
  2. RegisterNamespace: `namespaceId` -> `id`
  3. MosaicDefinition: `mosaicId` -> `id`
  4. HashLock: `mosaic` -> 'mosaicId`, Added `amount`
  5. SecretLock:  `mosaic` -> 'mosaicId`, `recipient` -> 'recipientAddress`, Added `amount`
  6. SecretProof: `recipient` -> 'recipientAddress`
  7. Transfer: `recipient` -> `recipientAddress`

Issue: #83

### CODE REVIEW

File: catapult-sdk/src/plugins/lockHash.js [hunk 1]
```diff
@@ -35,15 +35,16 @@ const lockHashPlugin = {
...
+			senderAddress: ModelType.binary,
 			mosaicId: ModelType.uint64,
 			height: ModelType.uint64,
 			hash: ModelType.binary
 		});
 
 		builder.addTransactionSupport(EntityType.hashLock, {
-			mosaic: ModelType.uint64,
+			mosaicId: ModelType.uint64,
+			amount: ModelType.uint64,
```
Comment (Vektrat) [inline catapult-sdk/src/plugins/lockHash.js]: why `amount`? it's not in catbuffer? https://github.com/nemtech/catbuffer/blob/master/schemas/lock_hash/hash_lock.cats

File: catapult-sdk/src/plugins/lockSecret.js [hunk 2]
```diff
@@ -35,23 +35,25 @@ const lockSecretPlugin = {
...
 			height: ModelType.uint64,
 			secret: ModelType.binary,
-			recipient: ModelType.binary
+			recipientAddress: ModelType.binary
 		});
 
 		builder.addTransactionSupport(EntityType.secretLock, {
 			mosaicId: ModelType.uint64,
 			duration: ModelType.uint64,
+			amount: ModelType.uint64,
```
Comment (Vektrat) [inline catapult-sdk/src/plugins/lockSecret.js]: why `amount`? it's not in catbuffer: https://github.com/nemtech/catbuffer/blob/master/schemas/lock_secret/secret_lock.cats

File: catapult-sdk/src/plugins/lockSecret.js [hunk 3]
```diff
@@ -35,23 +35,25 @@ const lockSecretPlugin = {
...
+			hashAlgorithm: ModelType.uint8,
```
Comment (Vektrat) [inline catapult-sdk/src/plugins/lockSecret.js]: `hashAlgorithm` should not be here because it is a basic type (uint8, remove line)

File: catapult-sdk/src/plugins/mosaic.js [hunk 4]
```diff
@@ -44,16 +44,20 @@ const mosaicPlugin = {
...
+			ownerPublicKey: ModelType.binary,
+			ownerAddress: ModelType.binary,
+			revision: 'number',
+			properties: { type: ModelType.object, schemaName: 'mosaicDefinition.mosaicProperty' }
 		});
 
 		builder.addSchema('mosaicDefinition.mosaicProperty', {
-			value: ModelType.uint64
+			flags: ModelType.uint8,
+			divisibility: ModelType.uint8,
```
Comment (Vektrat) [inline catapult-sdk/src/plugins/mosaic.js]: `flags` and `divisibility` should be removed (basic types)

File: catapult-sdk/src/plugins/mosaic.js [hunk 5]
```diff
@@ -44,16 +44,20 @@ const mosaicPlugin = {
...
-			owner: ModelType.binary,
-			properties: { type: ModelType.array, schemaName: 'mosaicDefinition.mosaicProperty' }
+			startHeight: ModelType.uint64,
+			ownerPublicKey: ModelType.binary,
+			ownerAddress: ModelType.binary,
+			revision: 'number',
+			properties: { type: ModelType.object, schemaName: 'mosaicDefinition.mosaicProperty' }
 		});
 
 		builder.addSchema('mosaicDefinition.mosaicProperty', {
```
Comment (Vektrat) [inline catapult-sdk/src/plugins/mosaic.js]: rename `mosaicDefinition.mosaicProperty` to `mosaicDefinition.mosaicProperties`

File: catapult-sdk/test/modelBinary/blockHeaderCodec_spec.js [hunk 6]
```diff
@@ -39,7 +39,7 @@ describe('block header codec', () => {
 				timestamp: [0x8B463A30, 0x54602D15],
 				difficulty: [0x30750286, 0xE87850E8],
 				previousBlockHash: Previous_Block_Hash_Buffer,
-				blockTransactionsHash: Block_Transactions_Hash_Buffer
+				transactionsHash: Block_Transactions_Hash_Buffer
```
Comment (Vektrat) [inline catapult-sdk/test/modelBinary/blockHeaderCodec_spec.js]: rename `Block_Transactions_Hash_Buffer` to match object property renaming

File: catapult-sdk/test/modelBinary/embeddedEntityCodec_spec.js [hunk 7]
```diff
@@ -23,15 +23,15 @@ const test = require('../binaryTestUtils');
...
+		const Signer_Buffer = Buffer.from(test.random.bytes(test.constants.sizes.signerPublicKey));
 
 		return {
 			buffer: Buffer.concat([
 				Signer_Buffer,
 				Buffer.of(0x2A, 0x81, 0x1C, 0x45) // version, type
 			]),
 			object: {
-				signer: Signer_Buffer,
+				signerPublicKey: Signer_Buffer,
```
Comment (Vektrat) [inline catapult-sdk/test/modelBinary/embeddedEntityCodec_spec.js]: rename `Signed_Buffer` to match new naming

File: catapult-sdk/test/modelBinary/verifiableEntityCodec_spec.js [hunk 8]
```diff
@@ -24,7 +24,7 @@ const test = require('../binaryTestUtils');
 describe('verifiable entity codec', () => {
 	const generateVerifiableEntity = () => {
 		const Signature_Buffer = Buffer.from(test.random.bytes(test.constants.sizes.signature));
-		const Signer_Buffer = Buffer.from(test.random.bytes(test.constants.sizes.signer));
+		const Signer_Buffer = Buffer.from(test.random.bytes(test.constants.sizes.signerPublicKey));
```
Comment (Vektrat) [inline catapult-sdk/test/modelBinary/verifiableEntityCodec_spec.js]: rename `SignedBuffer` to match renaming

File: catapult-sdk/test/plugins/aggregate_spec.js [hunk 9]
```diff
@@ -130,7 +130,7 @@ describe('aggregate plugin', () => {
 			const type = (options || {}).type || constants.knownTxType;
 			const extraSize = (options || {}).extraSize || 0;
 
-			const Signer_Buffer = Buffer.from(test.random.bytes(test.constants.sizes.signer));
+			const Signer_Buffer = Buffer.from(test.random.bytes(test.constants.sizes.signerPublicKey));
```
Comment (Vektrat) [inline catapult-sdk/test/plugins/aggregate_spec.js]: rename variable

File: catapult-sdk/test/plugins/transfer_spec.js [hunk 10]
```diff
@@ -88,7 +88,7 @@ describe('transfer plugin', () => {
 				]),
 
 				object: {
-					recipient: Recipient_Buffer
+					recipientAddress: Recipient_Buffer
```
Comment (Vektrat) [inline catapult-sdk/test/plugins/transfer_spec.js]: rename `Recipient_Buffer`

File: catapult-sdk/src/model/ModelType.js [hunk 11]
```diff
@@ -42,8 +42,11 @@ const ModelType = {
 	/** Schema property type indicating a uint16. */
 	uint16: SchemaType.max + 5,
 
+	/** Schema property type indicating a uint32 value. */
+	uint32: SchemaType.max + 6,
```
Comment (Vektrat) [inline catapult-sdk/src/model/ModelType.js]: why was this added? any specific reason? it was a basic type, so it was not needed, i don't see any changes made to the formatting rules

File: rest/src/plugins/mosaic/MosaicDb.js [hunk 12]
```diff
@@ -40,12 +40,12 @@ class MosaicDb {
...
 	mosaicsByIds(ids) {
 		const mosaicIds = ids.map(id => new Long(id[0], id[1]));
-		const conditions = { 'mosaic.mosaicId': { $in: mosaicIds } };
+		const conditions = { 'mosaic.id': { $in: mosaicIds } };
 		const collection = this.catapultDb.database.collection('mosaics');
 		return collection.find(conditions)
 			.sort({ _id: -1 })
 			.toArray()
-			.then(entities => Promise.resolve(this.catapultDb.sanitizer.copyAndDeleteIds(entities)));
+			.then(entities => Promise.resolve(this.catapultDb.sanitizer.deleteIds(entities)));
```
Comment (Vektrat) [inline rest/src/plugins/mosaic/MosaicDb.js]: why this change?

File: catapult-sdk/src/plugins/lockSecret.js (hunk 2 above)
Comment (rg911) [reply catapult-sdk/src/plugins/lockSecret.js]: `Amount` is in mongoDB. but I think you are right, it's removed.

File: catapult-sdk/src/plugins/lockHash.js (hunk 1 above)
Comment (rg911) [reply catapult-sdk/src/plugins/lockHash.js]: `Amount` is in mongoDB. but I think you are right, it's removed.

File: rest/src/plugins/mosaic/MosaicDb.js (hunk 12 above)
Comment (rg911) [reply rest/src/plugins/mosaic/MosaicDb.js]: `meta` does not exists anymore in `mosaics` collection

File: catapult-sdk/src/model/ModelType.js (hunk 11 above)
Comment (rg911) [reply catapult-sdk/src/model/ModelType.js]: that's right, will remove

File: catapult-sdk/src/plugins/lockHash.js (hunk 1 above)
Comment (Vektrat) [reply catapult-sdk/src/plugins/lockHash.js]: need to add amount to codecs (below) - note that ordering matters since codecs serialize and deserialize binary transaction data (must be put after `mosaic`)

File: catapult-sdk/src/plugins/lockSecret.js (hunk 2 above)
Comment (Vektrat) [reply catapult-sdk/src/plugins/lockSecret.js]: need to add amount to codecs (below) - note that ordering matters since codecs serialize and deserialize binary transaction data (must be put after `mosaic`)

File: catapult-sdk/src/plugins/lockHash.js (hunk 1 above)
Comment (rg911) [reply catapult-sdk/src/plugins/lockHash.js]: Fixed

File: catapult-sdk/src/plugins/lockSecret.js (hunk 2 above)
Comment (rg911) [reply catapult-sdk/src/plugins/lockSecret.js]: Fixed
